controller/api/public/edges.go

```diff
@@ -0,0 +1,144 @@
+package public
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	pb "github.com/linkerd/linkerd2/controller/gen/public"
+	"github.com/prometheus/common/model"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	inboundIdentityQuery  = "count(response_total%s) by (%s, client_id)"
+	outboundIdentityQuery = "count(response_total%s) by (%s, dst_%s, server_id, no_tls_reason)"
+)
+
+var formatMsg = map[string]string{
+	"disabled":                          "Disabled",
+	"loopback":                          "Loopback",
+	"no_authority_in_http_request":      "No Authority In HTTP Request",
+	"not_http":                          "Not HTTP",
+	"not_provided_by_remote":            "Not Provided By Remote",
+	"not_provided_by_service_discovery": "Not Provided By Service Discovery",
+}
+
+func (s *grpcServer) Edges(ctx context.Context, req *pb.EdgesRequest) (*pb.EdgesResponse, error) {
+	log.Debugf("Edges request: %+v", req)
+	if req.GetSelector().GetResource() == nil {
+		return edgesError(req, "Edges request missing Selector Resource"), nil
+	}
+
+	edges, err := s.getEdges(ctx, req)
+	if err != nil {
+		return edgesError(req, err.Error()), nil
+	}
+
+	return &pb.EdgesResponse{
+		Response: &pb.EdgesResponse_Ok_{
+			Ok: &pb.EdgesResponse_Ok{
+				Edges: edges,
+			},
+		},
+	}, nil
+}
+
+func edgesError(req *pb.EdgesRequest, message string) *pb.EdgesResponse {
+	return &pb.EdgesResponse{
+		Response: &pb.EdgesResponse_Error{
+			Error: &pb.ResourceError{
+				Resource: req.GetSelector().GetResource(),
+				Error:    message,
+			},
+		},
+	}
+}
+
+func (s *grpcServer) getEdges(ctx context.Context, req *pb.EdgesRequest) ([]*pb.Edge, error) {
+	labelNames := promGroupByLabelNames(req.Selector.Resource)
+	if len(labelNames) != 2 {
+		return nil, errors.New("unexpected resource selector")
+	}
+	resourceType := string(labelNames[1]) // skipping first name which is always namespace
+	labels := promQueryLabels(req.Selector.Resource)
+	labelsOutbound := labels.Merge(promDirectionLabels("outbound"))
+	labelsInbound := labels.Merge(promDirectionLabels("inbound"))
+
+	inboundQuery := fmt.Sprintf(inboundIdentityQuery, labelsInbound, resourceType)
+	outboundQuery := fmt.Sprintf(outboundIdentityQuery, labelsOutbound, resourceType, resourceType)
+
+	inboundResult, err := s.queryProm(ctx, inboundQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	outboundResult, err := s.queryProm(ctx, outboundQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	edge := processEdgeMetrics(inboundResult, outboundResult, resourceType)
+	return edge, nil
+}
+
+func processEdgeMetrics(inbound, outbound model.Vector, resourceType string) []*pb.Edge {
```

Comment:
this code is super nice, i wonder if we could leverage it for stats down the road. 👍